src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,419 @@
+
+ Outputs:
+
+ Purpose: Store base_name and expr so they will be declared as
+          inputs when declare_recorded_inputs() is called
+
+\*******************************************************************/
+
+void input_recordert::record_input(
+  const irep_idt &base_name,
+  const exprt &expr)
+{
+  inputs.emplace_back(base_name, expr);
+}
+
+/*******************************************************************\
+
+Function: input_recordert::declare_recorded_inputs
+
+ Inputs:
+  init_code - The code block to add the code to
+
+ Outputs:
+
+ Purpose: Output code declaring all the recorded inputs as inputs
+
+\*******************************************************************/
+
+void input_recordert::declare_recorded_inputs(code_blockt &init_code)
+{
+  for(const input_datat &input : inputs)
+  {
+    codet input_code(ID_input);
+    input_code.operands().resize(2);
+    input_code.op0()=
+      address_of_exprt(index_exprt(
+        string_constantt(input.base_name),
+        from_integer(0, index_type())));
+    input_code.op1()=input.expr;
+    input_code.add_source_location()=location;
+    init_code.add(input_code);
+  }
+  inputs.clear();
+}
+
+class symbol_factoryt
+{
+  bool assume_non_null;
+  symbol_tablet &symbol_table;
+  input_recordert &input_recorder;
+  const source_locationt &loc;
+  namespacet ns;
+
+public:
+  symbol_factoryt(
+    bool _assume_non_null,
+    symbol_tablet &_symbol_table,
+    input_recordert &input_recorder,
+    const source_locationt &loc):
+      assume_non_null(_assume_non_null),
+      symbol_table(_symbol_table),
+      input_recorder(input_recorder),
+      loc(loc),
+      ns(_symbol_table)
+  {}
+
+  exprt allocate_object(
+    code_blockt &init_code,
+    const exprt &target_expr,
+    const typet &allocate_type,
+    const bool static_lifetime);
+
+  void gen_nondet_init(
+    code_blockt &init_code,
+    const exprt &expr);
+};
+
+/*******************************************************************\
+
+Function: symbol_factoryt::allocate_object
+
+ Inputs:
+  code - The code block to add the code to
+  target_expr - The expression which we are allocating a symbol for
+  allocate_type - The type to use for the symbol. If this doesn't match
+                  target_expr then a cast will be used for the assignment.
+
+ Outputs: Returns the address of the allocated symbol
+
+ Purpose: Create a symbol for a pointer to point to
+
+\*******************************************************************/
+
+exprt symbol_factoryt::allocate_object(
+  code_blockt &code,
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const bool static_lifetime)
+{
+  // Add the following code to init_code:
+  //   <target_expr> = &tmp$<temporary_counter>
+  // Store the following code in the input recorder:
+  //   INPUT("tmp$<temporary_counter>", tmp$<temporary_counter>);
+
+  symbolt &aux_symbol=c_new_tmp_symbol(symbol_table, loc, allocate_type);
+  aux_symbol.is_static_lifetime=static_lifetime;
```

Comment:
Oops, that was a mistake. Well, hopefully I'll stop using static lifetime tmp symbols, but if I do continue using them I'll make it an option, and definitely only do it in one place.